File: be/src/main/java/sg/flow/controllers/AccountController.java
package sg.flow.controllers;

import java.util.List;

import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;

import lombok.RequiredArgsConstructor;
import sg.flow.models.account.AccountWithTransactionHistory;
import sg.flow.models.account.BriefAccount;
import sg.flow.services.AccountServices.AccountService;

@Controller
@RequiredArgsConstructor
public class AccountController {

    private final String ACCOUNTS = "/accounts";

    private final AccountService accountService;

    @GetMapping(ACCOUNTS + "/getAccounts")
    public ResponseEntity<List<BriefAccount>> getAccounts(
            @AuthenticationPrincipal(expression = "userId") Integer userId) {
        return ResponseEntity.ok(accountService.getBriefAccounts(userId));
    }

    @GetMapping(ACCOUNTS + "/getAccountsWithTransactionHistory")
    public ResponseEntity<List<AccountWithTransactionHistory>> getAccountsWithTransactionHistory(
            @AuthenticationPrincipal(expression = "userId") Integer userId) {
        return ResponseEntity.ok(accountService.getAccountWithTransactionHistory(userId));
    }

    @GetMapping(ACCOUNTS + "/getAccount")
    public ResponseEntity<BriefAccount> getAccount(@AuthenticationPrincipal(expression = "userId") Integer userId,
            @RequestParam(name = "accountId") Long accountId) {
        BriefAccount briefAccount;

        try {
            briefAccount = accountService.getBriefAccount(userId, accountId);
        } catch (IllegalArgumentException e) {
            if (e.getMessage().equals("Account does not belong to user")) {
                return ResponseEntity.status(HttpStatusCode.valueOf(403)).build();
            }
            return ResponseEntity.status(HttpStatusCode.valueOf(500)).build();
        }

        if (briefAccount == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(briefAccount);
    }

    @GetMapping(ACCOUNTS + "/getAccountWithTransactionHistory")
    public ResponseEntity<AccountWithTransactionHistory> getAccountWithTransactionHistory(
            @AuthenticationPrincipal(expression = "userId") Integer userId,
            @RequestParam(name = "accountId") Long accountId) {
        AccountWithTransactionHistory accountWithTransactionHistory;

        try {
            accountWithTransactionHistory = accountService.getAccountWithTransactionHistory(userId, accountId);
        } catch (IllegalArgumentException e) {
            if (e.getMessage().equals("Account does not belong to user")) {
                return ResponseEntity.status(HttpStatusCode.valueOf(403)).build();
            }
            return ResponseEntity.status(HttpStatusCode.valueOf(500)).build();
        }

        if (accountWithTransactionHistory == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(accountWithTransactionHistory);
    }

}

File: be/src/main/kotlin/sg/flow/controllers/AccountController.kt
package sg.flow.controllers

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import sg.flow.models.account.AccountWithTransactionHistory
import sg.flow.models.account.BriefAccount
import sg.flow.services.AccountServices.AccountService

@RestController
@RequestMapping("/account")
class AccountController(private val accountService: AccountService) {

    @GetMapping("/getAccounts")
    suspend fun getAccounts(
            @AuthenticationPrincipal(expression = "userId") userId: Int
    ): ResponseEntity<List<BriefAccount>> {
        var briefAccounts: List<BriefAccount> = accountService.getBriefAccounts(userId)
        return ResponseEntity.ok(briefAccounts)
    }

    @GetMapping("/getAccountsWithTransactionHistory")
    suspend fun getAccountsWithTransactionHistory(
            @AuthenticationPrincipal(expression = "userId") userId: Int
    ): ResponseEntity<List<AccountWithTransactionHistory>> =
            ResponseEntity.ok(accountService.getAccountWithTransactionHistory(userId))

    @GetMapping("/getAccount")
    suspend fun getAccount(
            @AuthenticationPrincipal(expression = "userId") userId: Int,
            @RequestParam(name = "accountId") accountId: Long
    ): ResponseEntity<BriefAccount> {
        return try {
            val briefAccount = accountService.getBriefAccount(userId, accountId)
            ResponseEntity.ok(briefAccount)
        } catch (e: IllegalArgumentException) {
            when (e.message) {
                "Account does not belong to user" ->
                        ResponseEntity.status(HttpStatus.FORBIDDEN).build()
                "Account not found" -> ResponseEntity.notFound().build()
                else -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build()
            }
        }
    }

    @GetMapping("/getAccountWithTransactionHistory")
    suspend fun getAccountWithTransactionHistory(
            @AuthenticationPrincipal(expression = "userId") userId: Int,
            @RequestParam(name = "accountId") accountId: Long
    ): ResponseEntity<AccountWithTransactionHistory> {
        return try {
            val accountWithTransactionHistory =
                    accountService.getAccountWithTransactionHistory(userId, accountId)
            ResponseEntity.ok(accountWithTransactionHistory)
        } catch (e: IllegalArgumentException) {
            when (e.message) {
                "Account does not belong to user" ->
                        ResponseEntity.status(HttpStatus.FORBIDDEN).build()
                "Account not found" -> ResponseEntity.notFound().build()
                else -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build()
            }
        }
    }
}
